File: src/main/kotlin/com/castab/resend/email/Delivery.kt
package com.castab.resend.email

import com.castab.resend.domain.EmailDeliveryState
import java.time.OffsetDateTime

val DELIVERY_EVENT_TYPES = listOf(
    "email.delivered",
    "email.delivery_delayed",
    "email.bounced",
    "email.complained",
    "email.suppressed",
    "email.failed",
)

private val DELIVERY_STATE_BY_EVENT_TYPE = mapOf(
    "email.delivered" to EmailDeliveryState.DELIVERED,
    "email.delivery_delayed" to EmailDeliveryState.DELIVERY_DELAYED,
    "email.bounced" to EmailDeliveryState.BOUNCED,
    "email.complained" to EmailDeliveryState.COMPLAINED,
    "email.suppressed" to EmailDeliveryState.SUPPRESSED,
    "email.failed" to EmailDeliveryState.FAILED,
)

fun isDeliveryEventType(type: String): Boolean = DELIVERY_STATE_BY_EVENT_TYPE.containsKey(type)

data class StoredDeliveryEvent(
    val id: String,
    val eventType: String,
    val eventCreatedAt: OffsetDateTime,
    val webhookReceivedAt: OffsetDateTime,
    val deliveryDetail: String?,
)

data class ProjectedDeliveryState(
    val state: EmailDeliveryState,
    val detail: String?,
    val eventCreatedAt: OffsetDateTime,
)

data class DeliveryProjection(val current: ProjectedDeliveryState?, val deliveredAt: OffsetDateTime?)

private fun earlier(left: OffsetDateTime?, right: OffsetDateTime): OffsetDateTime =
    if (left != null && !left.isAfter(right)) left else right

/**
 * Reduces the ordered stored delivery events into the current projected state, matching the Express
 * `reduceDeliveryEvents`: `delivery_delayed` never overwrites a non-delayed state, and `delivered`
 * tracks the earliest delivered timestamp.
 */
fun reduceDeliveryEvents(events: List<StoredDeliveryEvent>): DeliveryProjection {
    var deliveredAt: OffsetDateTime? = null
    var current: ProjectedDeliveryState? = null
    for (event in events) {
        val state = DELIVERY_STATE_BY_EVENT_TYPE[event.eventType] ?: continue
        if (state == EmailDeliveryState.DELIVERED) {
            deliveredAt = earlier(deliveredAt, event.eventCreatedAt)
        }
        if (state == EmailDeliveryState.DELIVERY_DELAYED &&
            current != null &&
            current.state != EmailDeliveryState.DELIVERY_DELAYED
        ) {
            continue
        }
        current = ProjectedDeliveryState(state, event.deliveryDetail, event.eventCreatedAt)
    }
    return DeliveryProjection(current, deliveredAt)
}

File: src/lib/email/delivery.ts
import type { Prisma, PrismaClient } from '@/lib/database';
import type { EmailWebhookEvent } from './webhook';

export const DELIVERY_EVENT_TYPES = [
  'email.delivered',
  'email.delivery_delayed',
  'email.bounced',
  'email.complained',
  'email.suppressed',
  'email.failed',
] as const;

type DeliveryEventType = (typeof DELIVERY_EVENT_TYPES)[number];
type DeliveryState =
  | 'DELIVERED'
  | 'DELIVERY_DELAYED'
  | 'BOUNCED'
  | 'COMPLAINED'
  | 'SUPPRESSED'
  | 'FAILED';
type DeliveryClient = PrismaClient | Prisma.TransactionClient;

interface StoredDeliveryEvent {
  id: string;
  eventType: string;
  eventCreatedAt: Date;
  webhookReceivedAt: Date;
  deliveryDetail: string | null;
}

const DELIVERY_STATE_BY_EVENT_TYPE: Record<DeliveryEventType, DeliveryState> = {
  'email.delivered': 'DELIVERED',
  'email.delivery_delayed': 'DELIVERY_DELAYED',
  'email.bounced': 'BOUNCED',
  'email.complained': 'COMPLAINED',
  'email.suppressed': 'SUPPRESSED',
  'email.failed': 'FAILED',
};

export function isDeliveryEventType(type: string): type is DeliveryEventType {
  return (DELIVERY_EVENT_TYPES as readonly string[]).includes(type);
}

export function getDeliveryDetail(event: EmailWebhookEvent): string | null {
  if (event.type === 'email.bounced') {
    return (
      event.data.bounce?.message ??
      event.data.bounce?.diagnosticCode?.join('\n') ??
      event.data.bounce?.type ??
      null
    );
  }
  if (event.type === 'email.failed') {
    return event.data.failed?.reason ?? null;
  }
  if (event.type === 'email.suppressed') {
    return (
      event.data.suppressed?.message ?? event.data.suppressed?.type ?? null
    );
  }
  return null;
}

export async function projectOutboundDeliveryStateForResendEmail(
  client: DeliveryClient,
  resendEmailId: string,
) {
  await client.$queryRaw`
    SELECT pg_advisory_xact_lock(hashtext(${resendEmailId}))::text
  `;

  const events = await client.emailWebhookEvent.findMany({
    where: {
      emailId: resendEmailId,
      eventType: { in: [...DELIVERY_EVENT_TYPES] },
    },
    orderBy: [
      { eventCreatedAt: 'asc' },
      { webhookReceivedAt: 'asc' },
      { id: 'asc' },
    ],
  });
  const projected = reduceDeliveryEvents(events);
  if (!projected.current) {
    return;
  }

  await client.emailMessage.updateMany({
    where: { direction: 'OUTBOUND', resendEmailId },
    data: {
      deliveryState: projected.current.state,
      deliveryStateDetail: projected.current.detail,
      deliveryStateUpdatedAt: projected.current.eventCreatedAt,
      deliveredAt: projected.deliveredAt,
    },
  });
}

export async function reconcileOutboundDeliveryState(
  client: DeliveryClient,
  resendEmailId: string,
) {
  await projectOutboundDeliveryStateForResendEmail(client, resendEmailId);
}

function reduceDeliveryEvents(events: StoredDeliveryEvent[]) {
  let deliveredAt: Date | null = null;
  let current: {
    state: DeliveryState;
    detail: string | null;
    eventCreatedAt: Date;
  } | null = null;

  for (const event of events) {
    if (!isDeliveryEventType(event.eventType)) {
      continue;
    }
    const state = DELIVERY_STATE_BY_EVENT_TYPE[event.eventType];
    if (state === 'DELIVERED') {
      deliveredAt = earlierDate(deliveredAt, event.eventCreatedAt);
    }
    if (
      state === 'DELIVERY_DELAYED' &&
      current &&
      current.state !== 'DELIVERY_DELAYED'
    ) {
      continue;
    }
    current = {
      state,
      detail: event.deliveryDetail,
      eventCreatedAt: event.eventCreatedAt,
    };
  }

  return { current, deliveredAt };
}

function earlierDate(left: Date | null, right: Date): Date {
  return left && left.getTime() <= right.getTime() ? left : right;
}
